ipfs-pinning-service.yaml

```
@@ -26,12 +26,23 @@ This section describes the most important object types and conventions.
 
 A full list of fields and schemas can be found in the `schemas` section of the [YAML file](https://github.com/ipfs/pinning-services-api-spec/blob/master/ipfs-pinning-service.yaml).
 
+
+## Identifiers
+
+### cid
+
+[Content Identifier (CID)](https://docs.ipfs.io/concepts/content-addressing/) points at the root of a DAG that is pinned recursively.
+
+### requestid
+
+Unique identifier of a pin request. Multiple request can point at the same `cid` but have different `name` or other metadata attached, `requestid` is used for differentiating between them.
```

```suggestion
Unique identifier of a pin request. When a pin is created, removed and then the same `cid` is pinned again the `requestid` is used for differentiating between those pin requests to avoid race conditions.
```

@obo20 Would this be a better explanation?

This explains how `requestid` helps with race conditions without being overly prescriptive about supporting more than one pin object for the same cid at a time.

---

@lidel to clarify, with this paradigm, is it currently up to the pinning service whether or not they want to allow concurrent duplicate pins of the same CID?

---

The change I proposed above enables greenfield services to support duplicate pins if they wish to do so, but given our timeline and available bandwidth I believe it is not a hard requirement.  

The idea here is that we keep the door open for implementing that capability, but if a service does not allow for duplicate pins, it can return an error that CID is already pinned by some other `requestid` and still be compliant with the spec. 

@jacobheun @aschmahmann @andrew thoughts?

---

For the @ipfs/wg-pinning-services people, It may help to provide some context here. @lidel and I spoke over zoom yesterday and discussed some of the complexities involved with allowing the same account to pin the same CID multiple times with different metadata and I described how right now this would be fairly challenging for us to accomplish due to the way our system handles accounting on a per-user-per-cid basis. (We would effectively need to create another layer of accounting / consensus on top of our existing layers).

I think this requirement is something we could eventually fulfill, but it would require a decent amount of rewriting how we handle internal pin accounting. With this in mind and the desire to ship with IPFS 0.8, it was questioned whether or not this requirement should be in the first "official" version of the pinning service spec.

To respond to your proposal @lidel, 

I think keeping things generic right now definitely provides the most flexibility for implementers when dealing with duplicates (an error can be returned if duplicates aren't supported). The biggest wildcard left to discuss is the "replace" endpoint. Unfortunately I think the endpoint is volatile enough that specific behaviors should be enforced. (Ideally behaviors that are easy to avoiding race conditions with).

---

Thanks! IIUC on the Pinata side, the resolution of this issue distills down to:

* When you want to create a pin via `POST /pins` or `POST /pins/{requestid}` but there is already a `Pin` object with the same `Pin.cid` , Pinata would  return `error.reason` = `DUPLICATE_PINS_NOT_SUPPORTED`  and in human-readable `error.details` suggest use of "replace" or "delete" endpoints instead.  
  This way pin "override" at Pinata will be explicit instead of implicit, and removing the surface for race conditions. 

@obo20 As for the "replace" endpoint, should we explicitly state that `PinStatus.requestid` is generated by `hash(JSON.stringify(Pin)+PinStatus.created)` or something like that? That would remove need for coming up with ID and protect implementer from racy footguns.

---

> should we explicitly state that PinStatus.requestid is generated by hash(JSON.stringify(Pin)+PinStatus.created) or something like that? That would remove need for coming up with ID and protect implementer from racy footguns.

My concern is less about the how the ID is generated, but rather the conditions under which a new one is generated under the "replace" endpoint. You had mentioned the possibility that if you only updated the CID then the endpoint wouldn't generate a new `requestid`, while updating the metadata would. I would argue for returning a new `requestid` every time (no matter what was changed) as a much safer alternative.

---

Those comments were based on idea that pinning services will add `Pin` to IPFS and use CID of that object as `requestid`, but I agree we should not assume every vendor will store pin objects on IPFS. Many will use RDBS and generate identifiers by other means. 

I agree, returning new `requestid ` every time will be safer, but don't want lock vendors out from storing `Pin` object on IPFS, if they choose to do so, so need to be careful about wording here.

@obo20 I'd like to  include this in the spec, but don't want people to guess what level "uniqueness" is expected. Would below be clear enough?

> Services can generate unique `requestid` by calculating `hash({accessToken},{Pin},{PinStatus.created})` or by other means that provide equally strong protection against race conditions.

---

@lidel is there a reason it needs to be a hashed id? From what I understand, as long as it's a unique id it should fulfill the same purpose. 

We use UUIDs for basically everything, so I would rather the method of generating `requestid` be up to the pinning service (as long as the id returned is always unique). If we go this route we could just mention as part of the spec that this `requestid` must always be unique and provide a few suggestions (such as your hashing method, UUID, etc) as to how that ID can be generated.

---

I updated this paragraph in 351f9b4 and it now includes UUID as a valid example of an opaque identifier.